src/rust/metric-forwarder/src/cloudwatch_send.rs

```diff
@@ -0,0 +1,177 @@
+use crate::cloudwatch_logs_parse::Stat;
+use crate::error::MetricForwarderError;
+use async_trait::async_trait;
+use futures::future;
+use log::warn;
+use rayon::prelude::*;
+use rusoto_cloudwatch::PutMetricDataError;
+use rusoto_cloudwatch::{CloudWatch, CloudWatchClient, MetricDatum, PutMetricDataInput};
+use rusoto_core::RusotoError;
+use statsd_parser;
+use statsd_parser::Metric;
+use std::collections::vec_deque;
+use std::collections::VecDeque;
+
+mod units {
+    // strings accepted by CloudWatch MetricDatum.unit
+    pub const COUNT: &'static str = "Count";
+    pub const MILLIS: &'static str = "Milliseconds";
+}
+
+type PutResult = Result<(), RusotoError<PutMetricDataError>>;
+
+#[async_trait]
+pub trait CloudWatchPutter {
+    // a subset of trait CloudWatch with the 1 function we want
+    async fn put_metric_data(&self, input: PutMetricDataInput) -> PutResult;
+}
+
+#[async_trait]
+impl<T> CloudWatchPutter for T
+where
+    T: CloudWatch + Sync + Send,
+{
+    async fn put_metric_data(&self, input: PutMetricDataInput) -> PutResult {
+        self.put_metric_data(input).await
+    }
+}
+
+// Hardcoded for now to defer making a decision on how we want to do this
+const CLOUDWATCH_NAMESPACE: &'static str = "grapl";
+
+pub async fn put_metric_data(
+    client: &impl CloudWatchPutter,
+    metrics: &[MetricDatum],
+) -> Result<(), MetricForwarderError> {
+    /*
+    Call Cloudwatch to insert metric data. Does batching on our behalf.
+    */
+
+    let chunks = metrics.chunks(20).map(|chunk| chunk.to_vec());
+    let put_requests = chunks.map(|data: Vec<MetricDatum>| PutMetricDataInput {
+        // TODO: should we do different namespaces for different services?
+        namespace: CLOUDWATCH_NAMESPACE.to_string(),
+        metric_data: data,
+    });
+
+    let request_futures = put_requests.map(|input| client.put_metric_data(input));
+    let responses: Vec<PutResult> = future::join_all(request_futures).await;
+
+    // TODO: retries
+    let failures: usize = responses.iter().filter(|resp| resp.is_err()).count();
+    match failures {
+        0 => Ok(()),
+        _ => Err(MetricForwarderError::PutMetricDataError(failures)),
+    }
```

Comment:
unsure if this should be 'Err if any failure' or just a warn

Reply:
I think it might make more sense to collect the errors into the Err itself, or at least one of them. Seems fine to return an Err here and then let the caller decide how to log/ handle it, or retry it, etc.